src/Raven.Server/Documents/SubscriptionState.cs

```diff
@@ -0,0 +1,117 @@
+        {
+            try
+            {
+                if (await _connectionInUse.WaitAsync(timeToWait) == false)
+                {
+                    switch (incomingConnection.Strategy)
+                    {
+                        // we try to connect, if the resource is occupied, we will throw an exception
+                        // this piece of code could have been upper, but we choose to have it here, for better readability
+                        case SubscriptionOpeningStrategy.WaitForFree:
+                            throw new TimeoutException();
+                        case SubscriptionOpeningStrategy.OpenIfFree:
+                            throw new SubscriptionInUseException(
+                                $"Subscription {incomingConnection.SubscriptionId} is occupied, connection cannot be opened");
+                        case SubscriptionOpeningStrategy.TakeOver:
+                            if (_currentConnection?.Strategy == SubscriptionOpeningStrategy.ForceAndKeep)
+                                throw  new SubscriptionInUseException(
+                                    $"Subscription {incomingConnection.SubscriptionId} is occupied by a ForceAndKeep connection, connectionId cannot be opened");
+
+                            _currentConnection.ConnectionException = new SubscriptionClosedException("Closed by Takeover");
+                            _currentConnection?.CancellationTokenSource.Cancel();
+                        
+                            throw new TimeoutException();
+                        case SubscriptionOpeningStrategy.ForceAndKeep:
+                            _currentConnection.ConnectionException = new SubscriptionClosedException("Closed by ForceAndKeep");
+                            _currentConnection?.CancellationTokenSource.Cancel();
+                        
+                            throw new TimeoutException();
+                        default:
+                            throw new InvalidOperationException("Unknown subscription open strategy: " +
+                                                                incomingConnection.Strategy);
+                    }
+                }
+            }
+            catch (SubscriptionException e)
+            {
+                incomingConnection.ConnectionException = e;
+
+                while (_rejectedConnections.Count > 10)
+                {
+                    SubscriptionConnection options;
+                    _rejectedConnections.TryDequeue(out options);
+                }
+
+                throw;
+            }
+
+            while (_recentConnections.Count > 10)
+            {
+                SubscriptionConnection options;
+                _recentConnections.TryDequeue(out options);
+            }
+
+            _connectionInUse.Reset();
+            _currentConnection = incomingConnection;
+            return new DisposableAction(() => {
+                _recentConnections.Enqueue(incomingConnection);
```

Comment:
why are you cleaning it up in the connect, and adding to `_recentConnections` on dispose?